src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumneManager.java

```diff
@@ -0,0 +1,173 @@
+   */
+  ExternalVolumeManager(
+      boolean isTestMode,
+      String role,
+      String clientName,
+      SnowflakeServiceClient snowflakeServiceClient) {
+    this.role = role;
+    this.clientName = clientName;
+    this.snowflakeServiceClient = snowflakeServiceClient;
+    this.externalVolumeMap = new ConcurrentHashMap<>();
+    try {
+      this.clientPrefix =
+          isTestMode
+              ? "testPrefix"
+              : this.snowflakeServiceClient
+                  .clientConfigure(new ClientConfigureRequest(role))
+                  .getClientPrefix();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.CLIENT_CONFIGURE_FAILURE, e.getMessage());
+    }
+  }
+
+  /**
+   * Given a fully qualified table name, return the target storage by looking up the table name
+   *
+   * @param fullyQualifiedTableName the target fully qualified table name
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage<T, ExternalVolumeLocation> getStorage(
+      String fullyQualifiedTableName) {
+    // Only one chunk per blob in Iceberg mode.
+    StreamingIngestStorage<T, ExternalVolumeLocation> stage =
+        this.externalVolumeMap.get(fullyQualifiedTableName);
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
+          String.format("No external volume found for table %s", fullyQualifiedTableName));
+    }
+
+    return stage;
+  }
+
+  /**
+   * Add a storage to the manager by looking up the table name from the open channel response
+   *
+   * @param dbName the database name
+   * @param schemaName the schema name
+   * @param tableName the table name
+   * @param fileLocationInfo response from open channel
+   */
+  @Override
+  public void addStorage(
+      String dbName, String schemaName, String tableName, FileLocationInfo fileLocationInfo) {
+    String fullyQualifiedTableName =
+        Utils.getFullyQualifiedTableName(dbName, schemaName, tableName);
+
+    try {
+      this.externalVolumeMap.put(
+          fullyQualifiedTableName,
+          new StreamingIngestStorage<T, ExternalVolumeLocation>(
+              this,
+              this.clientName,
+              fileLocationInfo,
+              new ExternalVolumeLocation(dbName, schemaName, tableName),
+              DEFAULT_MAX_UPLOAD_RETRIES));
+    } catch (SnowflakeSQLException | IOException err) {
+      throw new SFException(err, ErrorCode.UNABLE_TO_CONNECT_TO_STAGE);
+    }
+  }
+
+  /**
+   * Gets the latest file location info (with a renewed short-lived access token) for the specified
+   * location
+   *
+   * @param location A reference to the target location
+   * @param fileName optional filename for single-file signed URL fetch from server
+   * @return the new location information
+   */
+  @Override
+  public FileLocationInfo getRefreshedLocation(
+      ExternalVolumeLocation location, Optional<String> fileName) {
```

Comment:
Is this required for every file? I don't see this function being used

Reply:
Used [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/e808bbd6a215daaae9e5a49ec4006978afbeca8d/src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestStorage.java#L262) and [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/e808bbd6a215daaae9e5a49ec4006978afbeca8d/src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestStorage.java#L310).

Reply:
The content of this method will get replaced in a few days, we'll have a batch of file names with short lives tokens that the server sends down as part of channel configure (instead of sending the whole storage location's credentials which can lead to data corruption if someone overwrites another file in that folder with the provided token).
Once that happens, we will not have to call channelConfigure on every getRefreshedLocation call.